tls/extensions/s2n_cert_authorities.c

```diff
@@ -14,8 +14,77 @@
  */
 #include "tls/extensions/s2n_cert_authorities.h"
 
+#include <openssl/x509.h>
+
 #include "utils/s2n_safety.h"
 
+bool s2n_cert_authorities_supported_from_trust_store()
+{
+#if S2N_LIBCRYPTO_SUPPORTS_X509_STORE_LIST
+    return true;
+#else
+    return false;
+#endif
+}
+
+static S2N_RESULT s2n_cert_authorities_set_from_trust_store(struct s2n_config *config, uint16_t *count)
+{
+    RESULT_ENSURE_REF(config);
+    RESULT_ENSURE_REF(count);
+    *count = 0;
+
+    if (!config->trust_store.trust_store) {
+        return S2N_RESULT_OK;
+    }
+
+#if S2N_LIBCRYPTO_SUPPORTS_X509_STORE_LIST
+    DEFER_CLEANUP(struct s2n_stuffer output = { 0 }, s2n_stuffer_free);
+    RESULT_GUARD_POSIX(s2n_stuffer_growable_alloc(&output, 256));
+
+    STACK_OF(X509_OBJECT) *objects = X509_STORE_get0_objects(config->trust_store.trust_store);
+    RESULT_ENSURE(objects, S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+
+    int objects_count = sk_X509_OBJECT_num(objects);
+    RESULT_ENSURE(objects_count >= 0, S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+
+    for (int i = 0; i < objects_count; i++) {
+        X509 *cert = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(objects, i));
+        if (cert == NULL) {
+            /* X509_OBJECTs can also be CRLs, resulting in NULL here. Skip. */
+            continue;
+        }
+
+        X509_NAME *name = X509_get_subject_name(cert);
+        RESULT_ENSURE(name, S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+
+        const uint8_t *name_bytes = NULL;
+        size_t name_size = 0;
+        RESULT_GUARD_OSSL(X509_NAME_get0_der(name, &name_bytes, &name_size),
+                S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+
+        RESULT_ENSURE(*count < S2N_CERT_AUTHORITIES_MAX_COUNT, S2N_ERR_TOO_MANY_CAS);
+        RESULT_GUARD_POSIX(s2n_stuffer_write_uint16(&output, name_size));
+        RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&output, name_bytes, name_size));
+        RESULT_ENSURE(s2n_stuffer_data_available(&output) <= S2N_CERT_AUTHORITIES_MAX_SIZE,
+                S2N_ERR_TOO_MANY_CAS);
+        (*count)++;
+    }
+
+    RESULT_GUARD_POSIX(s2n_stuffer_extract_blob(&output, &config->cert_authorities));
+    return S2N_RESULT_OK;
+#else
+    RESULT_BAIL(S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
```

Comment:
I considered the existing S2N_ERR_NO_SUPPORTED_LIBCRYPTO_API / "libcrypto does not support this API", but that's an internal error. But maybe it's better anyway?

Or maybe I need a usage error S2N_ERR_FEATURE_DISABLED? An error specifically for this feature seems a little silly, since it's so niche. How long all the names I came up with are makes me think they're too specific xD

Reply:
I think it's because you can't usually get NO_SUPPORTED_LIBCRYPTO_API by calling an API. Mostly what's not supported by old libcryptos is newer algorithms, but we just disable those / skip them in preference lists, so customers never see the errors they'd get if they tried to use them.

Reply:
So this might be the first API that's disabled by using old libcrypto.